Introduce type common.ExtHash

blukat29 opened the pull request:
## Proposed changes

- Introduce ExtHash type for the state-pruning MPT node hashes.
- See [KIP-111](https://github.com/klaytn/kips/pull/112) for background. 

## Types of changes

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

## Further comments

aidan-kwon commented on common/types.go:
```diff
@@ -22,35 +22,53 @@ package common
+
+	extHashCounterMin  = uint64(0x100) // Any counter below extHashCounterMin are reserved.
+	extHashCounter     = extHashCounterMin
+	extHashRootNonce   = ExtHashNonce{0, 0, 0, 0, 0, 0x11}
+	extHashLegacyNonce = ExtHashNonce{0, 0, 0, 0, 0, 0x45}
```
@ethan-kr We need rationales for these numbers in KIP-112

ethan-kr replied:
I just needed a magic number to differentiate, and I put my personal favorite number.

aidan-kwon commented on common/types.go:
```diff
@@ -158,6 +176,142 @@ func (h UnprefixedHash) MarshalText() ([]byte, error) {
+// If len(b) is 0 to 32, then interpret b as Hash and then attaches LegacyNonce.
+// Otherwise b is interpreted as ExtHash. If len(b) is larger than len(ExtHash), b
+// will be cropped from the left.
+func BytesToExtHash(b []byte) ExtHash {
```
Is it okay this function doesn't check the length strictly? 
This function allows an input byte array having less than 32 lengths, then someone may use the leading zero removed byte array for hash. What if the array has 26 bytes for hash and  6 bytes for an extHashNonce? It will consider w

blukat29 replied:
Modified to strictly check if the length is 0, HashLength, or ExtHashLength.

blukat29 commented:
Update:
- rebased to latest dev
- strict length check in BytesToExtHash
- nonce is now 7 bytes

blukat29 commented:
I removed RootNonce which was deemed unnecessary in the KIP-111 implementation.
@ethan-kr @aidan-kwon PTAL.

ian0371 reviewed (commented):
I think we need to test BytesTo* functions separately (like TestBytesConversion)

blukat29 commented on common/types.go:
```diff
@@ -158,6 +171,147 @@ func (h UnprefixedHash) MarshalText() ([]byte, error) {
+	return eh.Unextend().getShardIndex(shardMask)
+}
+
+func EmptyExtHash(eh ExtHash) bool {
```
Users must use `EmptyExthash(eh)` or `(eh.Unextend() == common.Hash{})`.

In fact, the [empty hash + nonempty nonce] combination is very common in future PRs.
The ExtHash type sometimes appears in return values, but more often appears in function arguments.

In #1859, you'll see a lot of changes like below:

```diff
-     Foo(hash)
+     Foo(hash.ExtendLegacy())

- func Foo(hash common.Hash) {
-     if (hash == common.Hash{}) {
+ func Foo(hash common.ExtHash) {
+     if EmptyExtHash(hash) {
```

The `Hash` could have been zero, and `hash.ExtendLegacy()` ends with a nonzero nonce.
Therefore `EmptyExtHash` must be used to preserve the original logic.

ian0371 replied:
I see. so i guess there's no way to force `EmptyExtHash` and we should filter invalid checks in PR. Thanks

blukat29 replied:
Yes. Fortunately, changing the argument type to ExtHash causes compile errors so we won't miss such sites.

blukat29 commented on common/types.go:
```diff
@@ -158,6 +171,147 @@ func (h UnprefixedHash) MarshalText() ([]byte, error) {
+// If len(b) is 39, then b is interpreted as an ExtHash.
+// Otherwise, this function panics.
+func BytesToExtHash(b []byte) ExtHash {
+	if len(b) == 0 || len(b) == HashLength {
```
There is a rare case in NodeIterator where `BytesToExtHash(nil)` is called. I can't find the concrete call path right now, but it did happen while I work on #1859.

blukat29 replied:
The only sites I've discovered are in https://github.com/klaytn/klaytn/blob/v1.10.2/storage/statedb/iterator.go#L339-L375, where `common.BytesToHash(hash)` may be called with `hash == nil`. It's part of NodeIterator's valid operation, so it must be allowed.

blukat29 commented on common/types.go:
```diff
@@ -158,6 +171,147 @@ func (h UnprefixedHash) MarshalText() ([]byte, error) {
+	return h
+}
+
+// Nonce returns the 6 byte nonce component of an ExtHash
```
Fixed in 04a8c329.

blukat29 commented on common/types.go:
```diff
@@ -158,6 +171,147 @@ func (h UnprefixedHash) MarshalText() ([]byte, error) {
+	}
+}
+
+func BytesToExtNonce(b []byte) ExtHashNonce {
```
Fixed in 04a8c329.

blukat29 commented on common/types.go:
```diff
@@ -158,6 +171,147 @@ func (h UnprefixedHash) MarshalText() ([]byte, error) {
+}
+
+// ResetExtHashNonce sets the extHashCounter for deterministic testing
+func ResetExtHashCounter(counter uint64) {
```
Fixed in 04a8c329.

ian0371 replied:
name in the comment is not fixed

blukat29 replied:
Oops, fixed.

blukat29 commented:
> I think we need to test BytesTo* functions separately (like TestBytesConversion)

Added 050d2d83. Tell me if you are thinking of specific test cases other than this commit.

aidan-kwon commented on common/types.go:
```diff
@@ -22,35 +22,48 @@ package common
+
+	extHashCounterMin  = uint64(0x100) // Any counter below extHashCounterMin are reserved.
+	extHashCounter     = extHashCounterMin
+	extHashLegacyNonce = ExtHashNonce{0, 0, 0, 0, 0, 0, 0x45}
```
The value doesn't have any meaning, right? Please comment about this for the later changes

aidan-kwon replied:
@ethan-kr Can you explain the meaning of the value? How about changing this to all-zero?

blukat29 replied:
Added comments as to why we have nonzero number.

aidan-kwon reviewed (dismissed):
ExtHash requires many memory copies. Let's have a performance test about this in QA process

blukat29 commented on common/types.go:
```diff
@@ -158,6 +172,157 @@ func (h UnprefixedHash) MarshalText() ([]byte, error) {
+}
+
+func BytesToExtNonce(b []byte) ExtHashNonce {
+	if len(b) > ExtHashNonceLength {
```
It's used in nextExtHashNonce. Maybe should I make the length check strict?

blukat29 commented on common/types.go:
```diff
@@ -158,6 +172,157 @@ func (h UnprefixedHash) MarshalText() ([]byte, error) {
+	var h Hash
+	copy(h[:], eh[:HashLength])
+	return h
+}
```
No, your suggestion is more concise. Shall I change other functions to declare return values like this?

blukat29 commented on common/types.go:
```diff
@@ -158,6 +172,157 @@ func (h UnprefixedHash) MarshalText() ([]byte, error) {
+	num := atomic.AddUint64(&extHashCounter, 1)
+	bin := make([]byte, 8)
+	binary.BigEndian.PutUint64(bin, num)
+	return BytesToExtNonce(bin[1:8])
```
Fair point. I will also add a comment to describe how nonce is calculated.

blukat29 replied:
I changed as you've suggested in commit 5ab78722ccb79f0871df9ac53c790179514c730e, and it's somewhat non-intuitive as well.

blukat29 commented:
Cleaned up and squashed. In a conversation with @ethan-kr and @aidan-kwon, we decided:
- LegacyNonce is all-zero. If we ever need to change it later, we can do it because LegacyNonce only exists in memory and not persisted.
- extHashCounter increments by 1. Added comment explaining the logic.
- No need to compare the initial extHashCounter against a specific value. Only check for zero to make sure any timestamp has returned.

@jeongkyun-oh @ian0371 PTAL.
